SmingHub/Sming
# Partition tables

Implement class-based API for Sming with support for all architectures.
See #1676 for discussion.

Summary of changes:

* Create new `Storage` component and port low-level Esp32 partition support
* Use `gen_esp32part.py` tool as basis for new `hwconfig.py` tool. Integrate into build system. Add new targets.
* Implement a C++ API. `Partition::SubType::App` and `Partition::SubType::Data` are strong enums used to identify all the standard types, but the base API still uses type/subtype so custom types can be accommodated.
* Revise esp8266 sector layout - see updated documentation. This places all the critical information at the start of flash, whereas previously it was at the end. This avoids issues where flash size is set incorrectly.
* Custom `sysParam` data partition sub-type used for esp8266 system parameters and RF calibration data.
* Add support for 'external' devices (e.g. SPIRAM, supplementary flash, EEPROM, etc.) using a custom `Storage` partition type. Applications register a `Storage::Device` object implementation with the Partition API to register these external devices with a C++ interface for them.
* Disentangle SPIFFS from rBoot - OTA upgrading uses partition API.
* Drive SPIFFS image creation from partition table entries - more flexible, multiple images.
* Generate build variables directly from hardware config file (see `out/.../hwconfig.mk`). Preserve existing build variables where possible, document changes.
* Document hardware configuration with schema and incorporate validation. This can assist with fixing malformed configurations and supplements checks included in the partition tool.
* Add `Basic_Storage` sample to more thoroughly demonstrate usage
* Update documentation
* [x] Test (HostTests) on all arches
* [x] Remove `DISABLE_SPIFFS`, `SPI_SIZE` and `SPIFF_SIZE` build variables from all samples, including those in submodules.
* [x] Include samples and tests within submodules to CI
* [x] Add check to ensure image size will fit within partition

For other PR:

* Revise `rBoot` and `OtaUpgrade` to make more effective use of partitions.
* Update FlashString to support external data storage by Partition?
* Add NVS Component based on or using ESP-IDF for storing configuration data.
* Switch to spiffsgen.py (IFS)
